walkthrough/05-aws/02-stacks/README.md

```diff
@@ -0,0 +1,102 @@
+    └── db
+        ├── .terraform.lock.hcl
+        └── terragrunt.hcl
+```
+
+The only significant change that has taken place is the introduction of `terragrunt.stack.hcl` files in the `live` directories. These replace the tree of directories that were previously used to define the `api` and `db` Units.
+
+So far, this adjustment has cut down the total number of files in the project by 4, and has made the project more manageable, as updates to the `terragrunt.hcl` files used to define `api` and `db` Units are now defined in a single file. They are referenced by the `terragrunt.stack.hcl` files in the `dev` and `prod` environments within `live`.
+
+This chapter was authored by:
+1. Copying + pasting the previous one.
+2. Replacing the contents of `live/dev` and `live/prod` with `terragrunt.stack.hcl` files (while maintaining the `environment.hcl` files).
+3. Linking to the `mock-stack-generate.sh` script in the root of the project.
+
+### Quick Detour on `.terraform.lock.hcl` Files
+
+You may notice that the `units` directory contains directories that contain both `terragrunt.hcl` files and `.terraform.lock.hcl` files. These `.terraform.lock.hcl` files are generated by OpenTofu/Terraform when it initializes a module, and are used to lock the versions of the providers that are used by it. It can be useful to consistently use the same versions of providers, especially when working in a team, to avoid issues that can arise from different versions of providers behaving differently.
+
+You generally want to commit these files to your repository, as that ensures that updates go through git, but one of the incidental advantages to the design of Stacks is that you only need to track the `.terraform.lock.hcl` file once for the definition of a Unit instead of once per instance of a Unit.
+
+## Walkthrough
+
+To try out the version of this walkthrough that utilizes mock Stacks prototype, run the following commands:
+
+```bash
+$ cd live
+$ ./mock-stack-generate.sh
+$ terragrunt run-all apply --terragrunt-non-interactive
+```
+
+You'll notice that the resources are still provisioned, just as before, but that the Units are now nested in `.terragrunt-stack` directories:
+
+```bash
+$ terragrunt run-all apply --terragrunt-non-interactive
+14:58:27.656 INFO   The stack at /Users/yousif/repos/src/github.com/yhakbar/terragrunt-3313-stacks-walkthrough/walkthrough/05-aws/02-stacks/live will be processed in the following order for command apply:
+Group 1
+- Module dev/.terragrunt-stack/storage/db
+- Module prod/.terragrunt-stack/storage/db
+
+Group 2
+- Module dev/.terragrunt-stack/services/api
+- Module prod/.terragrunt-stack/services/api
+...
+```
+
+## Cleanup
+
+As before, you can clean up the resources by running:
+
+```bash
+terragrunt run-all destroy --terragrunt-non-interactive
+```
+
+## Next Steps
+
+The [next chapter](../03-no-shared/) will demonstrate how a little more refactoring can further reduce the amount of code that has to be maintained.
+
```

Comment:
The point at this stage is that the definition of the units are separately defined [here](https://github.com/yhakbar/terragrunt-3313-stacks-walkthrough/tree/main/walkthrough/05-aws/02-stacks/units) instead of repeatedly in each "Stack" as copied `terragrunt.hcl` files.

This way, if `dev` and `prod` reference the same Unit definition, there's only one place to edit Unit configurations instead of N places.

As you point out later, there is a DRY-er approach where the entire Stack is an encapsulated artifact, but users should be able to chose the level of abstraction that works for them.

Reply:
I actually love that ability to choose the abstraction level, but it wasn't clear to me that's an option here, so might be worth making that more explicit -- that here we're copying the stack definition, but that if you want to reuse the same dedinition, you'll show that later.